be/dtm0_log.c

```diff
@@ -0,0 +1,631 @@
+					       struct m0_dtm0_log_rec *rec)
+{
+	int rc;
+
+	/* TODO: dissolve dlr_txd and remove this code */
+	rc = m0_dtm0_tx_desc_copy(&rec->dlr_dtx.dd_txd, &rec->dlr_txd);
+	if (rc != 0)
+		return rc;
+
+	lrec_tlink_init_at_tail(rec, log->dl_tlist);
+	return 0;
+}
+
+M0_INTERNAL void m0_be_dtm0_volatile_log_update(struct m0_be_dtm0_log  *log,
+						struct m0_dtm0_log_rec *rec)
+{
+	/* TODO: dissolve dlr_txd and remove this code */
+	m0_dtm0_tx_desc_apply(&rec->dlr_txd, &rec->dlr_dtx.dd_txd);
+}
+
+M0_INTERNAL int m0_be_dtm0_plog_can_prune(struct m0_be_dtm0_log	    *log,
+					  const struct m0_dtm0_tid  *id,
+					  struct m0_be_tx_credit    *cred)
+{
+	/* This assignment is meaningful as it covers the empty log case */
+	int                     rc = M0_DTS_LT;
+	struct m0_dtm0_log_rec *rec;
+	struct m0_be_list      *dl_list = log->dl_list;
+
+	M0_PRE(m0_be_dtm0_log__invariant(log));
+	M0_PRE(m0_dtm0_tid__invariant(id));
+	M0_PRE(m0_mutex_is_locked(&log->dl_lock));
+
+	m0_be_list_for(lrec, dl_list, rec) {
+		if (!m0_dtm0_tx_desc_state_eq(&rec->dlr_txd,
+					      M0_DTPS_PERSISTENT))
+			return M0_ERR(-EPROTO);
+
+		if (cred) {
+			m0_be_dtm0_log_credit(M0_DTML_PRUNE, NULL, NULL,
+					      log->dl_seg, rec, cred);
+		}
+		rc = m0_dtm0_tid_cmp(log->dl_cs, &rec->dlr_txd.dtd_id, id);
+		if (rc != M0_DTS_LT)
+			break;
+	} m0_be_list_endfor;
+
+	if (rc != M0_DTS_EQ)
+		return -ENOENT;
+	return 0;
+}
+
+M0_INTERNAL int m0_be_dtm0_plog_prune(struct m0_be_dtm0_log    *log,
+                                     struct m0_be_tx          *tx,
+                                     const struct m0_dtm0_tid *id)
+{
+	struct m0_dtm0_log_rec *rec;
+	struct m0_dtm0_tid cur_id = {};
+
+	M0_PRE(m0_be_dtm0_log__invariant(log));
+	M0_PRE(m0_dtm0_tid__invariant(id));
+	M0_PRE(m0_mutex_is_locked(&log->dl_lock));
+
+	m0_be_list_for (lrec, log->dl_list, rec) {
+		cur_id = rec->dlr_txd.dtd_id;
+
+		lrec_be_list_del(log->dl_list, tx, rec);
+		lrec_be_tlink_destroy(rec, tx);
+		be_dtm0_plog_rec_fini(&rec, log, tx);
+		if (m0_dtm0_tid_cmp(log->dl_cs, &cur_id, id) == M0_DTS_EQ)
+			break;
+	} m0_be_list_endfor;
+
+	return 0;
+}
+
+M0_INTERNAL void m0_be_dtm0_log_pmsg_post(struct m0_be_dtm0_log *log,
+					  struct m0_fop         *fop)
+{
+	struct m0_dtm0_log_rec       *rec;
+	struct dtm0_req_fop          *req = m0_fop_data(fop);
+	const struct m0_dtm0_tx_desc *txd = &req->dtr_txr;
+	bool                          is_stable;
+
+	M0_PRE(log != NULL);
+	M0_PRE(!log->dl_is_persistent);
+	M0_PRE(fop->f_type == &dtm0_req_fop_fopt);
+	M0_PRE(m0_dtm0_tx_desc__invariant(txd));
+
+	M0_ENTRY();
+
+	m0_mutex_lock(&log->dl_lock);
+	rec = m0_be_dtm0_log_find(log, &txd->dtd_id);
+	/* TODO: We do not handle the case where a P msg is received before
+	 * the corresponding DTX enters INPROGRESS state.
+	 */
+	M0_ASSERT_INFO(rec != NULL, "Log record must be inserted into the log "
+		       "in m0_dtx0_close().");
+	is_stable = m0_dtm0_tx_desc_state_eq(&rec->dlr_txd,
```

Comment:
Can we rename "is_stable" to "is_persistent"?

Reply:
Hmm, OK.